### PR TITLE
Wait last version search to release growing safely

### PR DESCRIPTION
**Pull request by congqixia**
Wait last version search to release growing safely

Related to #17679

/kind bug

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

### CODE REVIEW

**Comment by xiaofan-luan** [comment]
is this ready to merge?

File: internal/querynode/shard_cluster.go [hunk 1]
```diff
@@ -210,10 +212,16 @@ func (sc *ShardCluster) addNode(evt nodeEvent) {
 		defer oldNode.client.Stop()
 	}
 
+	client, err := sc.nodeBuilder(evt.nodeID, evt.nodeAddr)
+	if err != nil {
+		// target shall be offline
+		log.Warn("failed to get shardQueryNode", zap.Error(err))
+		return
```
Comment (xiaofan-luan) [inline internal/querynode/shard_cluster.go:219]: The err exception is weird here.

More common way is to call something like GetOrCreateClient in session.go.

Each time we create one client, used the cache if anyone.

File: internal/querynode/shard_cluster.go [hunk 2]
```diff
@@ -128,12 +128,15 @@ type ShardCluster struct {
...
-	currentVersion *ShardClusterVersion           // current serving segment state version
+	mut      sync.RWMutex
+	leader   *shardNode           // shard leader node instance
+	nodes    map[int64]*shardNode // online nodes
+	segments SegmentsStatus       // shard segments
+
+	mutVersion sync.RWMutex
+	//versions       map[int64]*ShardClusterVersion // version id to version map
+	versions       sync.Map             // version id to version
+	currentVersion *ShardClusterVersion // current serving segment state version
```
Comment (xiaofan-luan) [inline internal/querynode/shard_cluster.go:139]: since version is already sync.map. what about using Atomic for current value?

Comment (congqixia) [reply internal/querynode/shard_cluster.go:139]: mutVersion needs to block get version's segment allocation

File: internal/querynode/shard_cluster.go (hunk 1 above)
Comment (congqixia) [reply internal/querynode/shard_cluster.go:219]: working on it.

File: internal/querynode/shard_cluster.go [hunk 3]
```diff
@@ -511,13 +520,13 @@ func (sc *ShardCluster) finishUsage(versionID int64) {
...
-	sc.mut.Lock()
-	defer sc.mut.Unlock()
 
-	version, ok := sc.versions[versionID]
+	v, ok := sc.versions.Load(versionID)
 	if ok {
+		version := v.(*ShardClusterVersion)
 		version.FinishUsage()
 	}
+	log.Warn("CQX finish usage done")
```
[the remark this answers is not in the document]
Comment (congqixia) [reply internal/querynode/shard_cluster.go:529]: working on it.

File: internal/querynode/shard_cluster.go (hunk 1 above)
Comment (congqixia) [reply internal/querynode/shard_cluster.go]: This part shall be separated into another pr.

**Pull request merged**